test/gtest/uct/ib/test_devx.cc

```diff
@@ -49,3 +51,216 @@ UCS_TEST_P(test_devx, dbrec)
+        EXPECT_EQ(init_export, nullptr != md()->umr.qp);
+        EXPECT_EQ(init_export, nullptr != md()->umr.cq);
+        if (!init_export) {
+            EXPECT_TRUE(ucs_list_is_empty(&md()->umr.mkey_pool));
+        }
+
+        EXPECT_EQ(init_import, nullptr != md()->umr.mkey_hash);
+    }
+
+    static void check_memh_export(uct_ib_mlx5_devx_mem_t *memh, bool is_umr) {
+        EXPECT_EQ(is_umr, UCT_IB_MLX5_MKEY_TAG_UMR ==
+                          (memh->exported_lkey & 0xff));
+        EXPECT_EQ(is_umr, nullptr != memh->exported_umr_mkey);
+        EXPECT_EQ(is_umr, nullptr == memh->cross_mr);
+    }
+};
+
+UCS_TEST_P(test_devx_umr_mkey, lazy_init)
+{
+    check_umr_init(false, false);
+}
+
+UCS_TEST_SKIP_COND_P(test_devx_umr_mkey, export_mkey_no_xgvmi, !check_xgvmi())
+{
+    uct_ib_mlx5_devx_mem_t *memh = create_memh(1024);
+
+    /* Drop XGVMI capability */
+    md()->flags &= ~UCT_IB_MLX5_MD_FLAG_INDIRECT_XGVMI;
+
+    ASSERT_UCS_OK(uct_ib_mlx5_devx_reg_exported_key(md(), memh));
+
+    /* UMR objects remains uninitialized, exported key is not UMR one */
+    check_umr_init(false, false);
+    check_memh_export(memh, false);
+    destroy_memh(memh);
+}
+
+UCS_TEST_SKIP_COND_P(test_devx_umr_mkey, export_mkey_xgvmi, !check_xgvmi())
+{
+    uct_ib_mlx5_devx_mem_t *memh = create_memh(1024);
+    ASSERT_UCS_OK(uct_ib_mlx5_devx_reg_exported_key(md(), memh));
+    /* XGVMI capability might be dropped by this point! */
+
+    /* UMR export transport objects must be initialized */
+    check_umr_init(true, false);
+    /* UMR mkey pool must be empty */
+    ASSERT_TRUE(ucs_list_is_empty(&md()->umr.mkey_pool));
+    /* Whether mkey is UMR depends on XGVMI capability */
+    check_memh_export(memh, check_xgvmi());
+
+    destroy_memh(memh);
+
+    /* After release of memh, associated UMR mkey is moved back to mkey pool */
+    ASSERT_EQ(check_xgvmi() ? 1 : 0, ucs_list_length(&md()->umr.mkey_pool));
+}
+
+UCS_TEST_SKIP_COND_P(test_devx_umr_mkey, export_mkey_xgvmi_reuse,
+                     !check_xgvmi_effective())
+{
+    const size_t MEMH_COUNT = 10;
+    uct_ib_mlx5_devx_mem_t *memh[MEMH_COUNT];
+
+    for (size_t i = 0; i < MEMH_COUNT; ++i) {
+        memh[i] = create_memh(1024);
+        ASSERT_UCS_OK(uct_ib_mlx5_devx_reg_exported_key(md(), memh[i]));
+        check_memh_export(memh[i], true);
+    }
+
+    /* UMR mkey pool must be empty */
+    ASSERT_TRUE(ucs_list_is_empty(&md()->umr.mkey_pool));
+
+    for (size_t i = 0; i < MEMH_COUNT; ++i) {
+        destroy_memh(memh[i]);
+    }
+
+    /* After release all allocated UMR keys go back to the mkey pool */
+    ASSERT_EQ(MEMH_COUNT, ucs_list_length(&md()->umr.mkey_pool));
+
+    /* Now reuse the keys for different handles of different size */
+    const size_t REUSE_COUNT = 5;
+    for (size_t i = 0; i < REUSE_COUNT; ++i) {
+        memh[i] = create_memh(2048);
+        ASSERT_UCS_OK(uct_ib_mlx5_devx_reg_exported_key(md(), memh[i]));
```

Comment:
can we check performance or somehow check there is no FW flow when reusing an existing key?
checking mkey_pool list length is too low level and indirect

Reply:
Good question!
Regarding checking FW flow - that's a classical case for mocking, we could easily do this if we used mocking library/framework (that we discussed some time ago). Then we would write something like:
`mock().expectNcalls(0, "uct_ib_mlx5_devx_obj_create");`
But without mocking framework I've no idea how we can check that.

wrt check performance - is it fast and reliable way to guarantee that FW code is not triggered? I don't think so..
So IMO the only fast and reliable way to check this requirement is to check the pool size